src/allocator.rs

```diff
@@ -545,6 +551,25 @@ impl Allocator {
             .map_err(|_| EvalErr(node, "atom is not a G2 point".to_string()))
     }
 
+    pub fn visit_node<'a, T>(
+        &'a self,
+        node: NodePtr,
+        visitor: impl FnOnce(&NodeVisitor<'a>) -> T,
+    ) -> T {
+        match node.node_type() {
+            (ObjectType::Bytes, idx) => {
```

Comment:
Is there any concern that a value that could be represented as a small atom sneaks into the wrong node type and the `visitor` then does the wrong thing?

Reply:
Have you considered having this function (or maybe a different one) request the passed node to be represented in a particular way if possible? Then it could catch & cast the standard atom being incorrectly not coded as a small atom.

Reply:
`visit_node_as_number`, `visit_node_as_u32`, etc.

Reply:
Another thing to note is that `Number` could be made an enum switching between u64 and bignum, which might simplify some things.

Reply:
I think it might be a good idea to have something like `visit_node_as_bytes()`. It might deduplicate some instances if `u32` -> `[u8]` conversions. I'd like to punt on it until there's a clearer need for it. I'm envisioning making `G1`, `G2` and `Number` their own node types as well, at which point this might become a lot more convenient.

Regarding using `u64` or `i64` for arithmetic on small numbers, that was one of my first attempts at speeding things up, two years ago or so. It has pretty far reaching consequences, but I expect the end-game to incorporate something like that.